sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/SdkRequestContext.java

```diff
@@ -0,0 +1,160 @@
+    }
+
+    /**
+     * Creates a new instance of {@link SdkRequestContext} with the given options.
+     * @param options The {@link RequestOptions} to be used.
+     */
+    SdkRequestContext(RequestOptions options) {
+        this(options, null, null);
+    }
+
+    /**
+     * Creates a new instance of {@link SdkRequestContext} with the given options.
+     * @param options The {@link RequestOptions} to be used.
+     * @return The {@link SdkRequestContext} from the given options.
+     */
+    public static SdkRequestContext from(RequestOptions options) {
+        if (options != null && options.getRequestContext() != null) {
+            return options.getRequestContext();
+        }
+
+        return new SdkRequestContext(options);
+    }
+
+    /**
+     * Get the {@link RequestOptions} corresponding to this request context.
+     * @return The {@link RequestOptions} corresponding to this request context.
+     */
+    public RequestOptions toRequestOptions() {
+        return requestOptions;
+    }
+
+    /**
+     * An empty {@link SdkRequestContext} used in situations where there is no request-specific
+     * configuration to pass into the request.
+     *
+     * @return The singleton instance of an empty {@link RequestOptions}.
+     */
+    public static SdkRequestContext none() {
+        return NONE;
+    }
+
+    /**
+     * Get the {@link InstrumentationContext} from the request context.
+     * @return The {@link InstrumentationContext} if present, otherwise null.
+     */
+    public InstrumentationContext getInstrumentationContext() {
+        return requestOptions.getInstrumentationContext();
+    }
+
+    /**
+     * Get the {@link ProgressReporter} from the request context.
+     *
+     * @return The {@link ProgressReporter} if present, otherwise null.
+     */
+    public ProgressReporter getProgressReporter() {
+        if (childProgressReporter != null) {
+            return childProgressReporter;
+        }
+
+        Object progressReporter = requestOptions.getMetadata(PROGRESS_REPORTER_KEY);
+        if (progressReporter instanceof ProgressReporter) {
+            return (ProgressReporter) progressReporter;
+        }
+
+        if (progressReporter != null) {
+            LOGGER.atVerbose()
+                .addKeyValue("actualType", progressReporter.getClass().getCanonicalName())
+                .addKeyValue("contextKey", PROGRESS_REPORTER_KEY)
+                .log("Unexpected object type. Ignoring it.");
+        }
+
+        return null;
+    }
+
+    /**
+     * Set the {@link ProgressReporter} to be used for this request context.
+     *
+     * @param progressReporter The {@link ProgressReporter} to be used.
+     * @return A new instance of {@link SdkRequestContext} with the given {@link ProgressReporter}.
+     */
+    public SdkRequestContext setProgressReporter(ProgressReporter progressReporter) {
+        return new SdkRequestContext(this.requestOptions, progressReporter, null);
```

Comment:
Generally, our `set` semantics updates the same instance. Here, we are creating a new instance and it may not be used as intended

```java
SdkRequestContext context = SdkRequestContext.from(requestOptions);
context.setInstrumentationContext(instrumentationContext); // returns a new instance different from context
context.setProgressReporter(progressReporter); // returns another new instance different from context and the one returned in the previous line
```

Reply:
Should we just have another factory method?
```java
public SdkRequestContext from(SdkRequestContext current, InstrumentationContext context);
```